parsing/parchmint-parser.js

```diff
@@ -118,21 +150,88 @@ class ParchmintParser {
      *
      * @since 1.0.0
      *
-     * @param {string} parchmint
+     * @param {string|object}   parchmint   A Parchmint file. If the parameter
+     *                                      is of type string, it will be parsed
+     *                                      by {@link JSON.parse} before
+     *                                      continuing.
      */
     constructor(parchmint = Validation.DEFAULT_STR_VALUE) {
         this.parchmint = parchmint;
 
         this.valid = true;
         this.idSet = new Set();
 
+        this.architecture = null;
+
         this.layers = [];
         this.components = new Map();
         this.connections = new Map();
         this.compFeatures = new Map();
         this.connFeatures = new Map();
+
+        this.schemaValidator = ajv.compile(schema);
     }
 
+    /**
+     * Parses, builds, and validates the Architecture from the given Parchmint.
+     *
+     * Sets the parser invalid if the Architecture is invalid.
+     *
+     * @since 1.0.0
+     *
+     * @param {string|object}   parchmint   A Parchmint file. If the parameter
+     *                                      is of type string, it will be parsed
+     *                                      by {@link JSON.parse} before
+     *                                      continuing. If this parameter if
+     *                                      left as its default, the parser's
+     *                                      parchmint field will be used.
+     * @returns {Object}    An Architecture object representing this Parchmint
+     *                      file, or null if the Parchmint itself is invalid.
+     */
+    parse(parchmint=null) {
+        let file = parchmint ? parchmint : this.parchmint;
+        let obj;
+
+        if (typeof file === 'string') {
+            obj = JSON.parse(file);
+        } else if (typeof file === 'object') {
+            obj = file;
+        } else {
+            this.valid = false;
+            console.log('Parser (FATAL ERROR): Parchmint file was not represented as a string, nor an' +
+                    ' object.\nAborting.');
+            return null;
+        }
+
+        if (!this.schemaValidator(obj)) {
+            this.valid = false;
+            console.log('Parser (FATAL ERROR): ' + ajv.errorsText(this.schemaValidator.errors) + '\nAborting.');
+            return null;
+        }
+
+        // The only required keys are "layers" and "name", so we need to check whether these keys exist before
+        // trying to parse them.
+        if (obj['features']) {
+            this.parseComponentFeatures(obj);
+        }
+        if (obj['components']) {
+            this.parseComponents(obj);
+        }
+        if (obj['features']) {
+            this.parseConnectionFeatures(obj);
+        }
+        if (obj['connections']) {
+            this.parseConnections(obj);
+        }
```

Comment:
The ```parseComponents``` and ```parseConnections``` build components and features by looking for their associated features in the ```compFeatures``` and ```connFeatures``` maps. Those maps are filled by ```parseComponentFeatures``` and ```parseConnectionFeatures```, so they must be called first.

Reply:
That makes sense, can you write a quick walkthrough of the steps that need to happen to fully parse the input and the general order they need to go in into the README? Since its essentially the inverse of the way that we parse in Inkwell I think it would be good to document it for people who modify it in the future.

Reply:
I wanted to merge this PR so I opened issue #69 to cover this.